merino-cache/src/memory.rs

```diff
@@ -16,13 +16,59 @@ use merino_suggest::{
 };
 use std::{
     borrow::Cow,
+    collections::HashMap,
     fmt::Debug,
     hash::Hash,
     sync::Arc,
     time::{Duration, Instant},
 };
 use tracing::Instrument;
 
+use arc_swap::ArcSwap;
+
+lazy_static! {
+    static ref LOCK_TABLE: ArcSwap<HashMap<String, Instant>> =
+        ArcSwap::from_pointee(HashMap::new());
+}
+
+impl LOCK_TABLE {
+    /// Check to see if there's any lock for a given key
+    fn is_locked(&self, key: &str) -> bool {
+        if let Some(lock_val) = self.load().get(key) {
+            return *lock_val > Instant::now();
+        }
+        false
+    }
+
+    /// Generate a lock for the given key and timeout
+    fn add_lock(&self, key: &str, lock_timeout: Duration) -> Instant {
+        let lock = Instant::now() + lock_timeout;
+        self.rcu(|table| {
+            let mut locked = HashMap::clone(table);
+            locked.insert(key.to_owned(), lock);
+            locked
+        });
+        lock
+    }
+
+    /// run func only if the lock we have matches what is registered for the key
+    fn update<F>(&self, key: &str, lock: Instant, mut func: F)
```

Comment:
This function also unlocks the lock after the update happens right? That sounds like something that should be mentioned in the doc string or the function name.